File: src-tauri/src/services/db_crypto.rs
//! Password-protected SQLite at rest.
//! Working copy is standard SQLite; durable form is AES-GCM encrypted `personal.db.enc`.

use crate::error::{AppError, AppResult};
use crate::services::crypto::{decrypt, encrypt, KEY_LEN};
use std::fs;
use std::path::{Path, PathBuf};

pub fn db_plain_path(data_dir: &Path) -> PathBuf {
    data_dir.join("personal.db")
}

pub fn db_enc_path(data_dir: &Path) -> PathBuf {
    data_dir.join("personal.db.enc")
}

/// Ensure a usable plaintext DB exists after unlock.
pub fn unlock_database_file(data_dir: &Path, db_key: &[u8; KEY_LEN]) -> AppResult<PathBuf> {
    let plain = db_plain_path(data_dir);
    let enc = db_enc_path(data_dir);

    // Crash recovery: keep working copy if it already exists.
    if plain.exists() {
        seal_database_file(data_dir, db_key)?;
        return Ok(plain);
    }

    if enc.exists() {
        let blob = fs::read(&enc)?;
        let data = decrypt(db_key, &blob)?;
        fs::write(&plain, data)?;
        return Ok(plain);
    }

    // Fresh install — empty file created by rusqlite open.
    Ok(plain)
}

/// Encrypt working DB to durable ciphertext and remove plaintext if possible.
pub fn seal_database_file(data_dir: &Path, db_key: &[u8; KEY_LEN]) -> AppResult<()> {
    let plain = db_plain_path(data_dir);
    let enc = db_enc_path(data_dir);
    if !plain.exists() {
        return Ok(());
    }
    // Checkpoint WAL into main file before reading bytes.
    let _ = crate::database::Database::checkpoint_file(&plain);
    let data = fs::read(&plain)?;
    let blob = encrypt(db_key, &data)?;
    fs::write(&enc, blob)?;
    Ok(())
}

pub fn replace_plain_db_from_bytes(
    data_dir: &Path,
    db_key: &[u8; KEY_LEN],
    sqlite_bytes: &[u8],
) -> AppResult<PathBuf> {
    let plain = db_plain_path(data_dir);
    if let Some(parent) = plain.parent() {
        fs::create_dir_all(parent)?;
    }
    fs::write(&plain, sqlite_bytes)?;
    seal_database_file(data_dir, db_key)?;
    Ok(plain)
}

pub fn read_plain_db_bytes(data_dir: &Path) -> AppResult<Vec<u8>> {
    let plain = db_plain_path(data_dir);
    if !plain.exists() {
        return Err(AppError::Other("数据库尚未创建".into()));
    }
    let _ = crate::database::Database::checkpoint_file(&plain);
    Ok(fs::read(&plain)?)
}

File: server/internal/core/dbcrypto.go
package core

import (
	"os"
	"path/filepath"

	"personal-os-server/internal/crypto"
)

func dbPlainPath(dataDir string) string { return filepath.Join(dataDir, "personal.db") }
func dbEncPath(dataDir string) string   { return filepath.Join(dataDir, "personal.db.enc") }

// unsealDatabaseFile ensures a usable plaintext personal.db exists after unlock.
// Mirrors db_crypto.rs unlock_database_file (crash recovery keeps existing plaintext).
func unsealDatabaseFile(dataDir string, dbKey [crypto.KeyLen]byte) error {
	plain := dbPlainPath(dataDir)
	enc := dbEncPath(dataDir)
	if _, err := os.Stat(plain); err == nil {
		return nil // crash recovery: reuse working copy
	}
	if data, err := os.ReadFile(enc); err == nil {
		pt, err := crypto.Decrypt(dbKey, data)
		if err != nil {
			return err
		}
		return os.WriteFile(plain, pt, 0o600)
	}
	return nil // fresh install; sqlite will create the file
}

// sealPlainFile encrypts the plaintext db bytes into personal.db.enc.
func sealPlainFile(dataDir string, dbKey [crypto.KeyLen]byte) error {
	plain := dbPlainPath(dataDir)
	data, err := os.ReadFile(plain)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	blob, err := crypto.Encrypt(dbKey, data)
	if err != nil {
		return err
	}
	return os.WriteFile(dbEncPath(dataDir), blob, 0o600)
}

// removePlainArtifacts deletes the plaintext DB and its WAL/SHM sidecars.
func removePlainArtifacts(dataDir string) {
	plain := dbPlainPath(dataDir)
	for _, p := range []string{plain, plain + "-wal", plain + "-shm"} {
		_ = os.Remove(p)
	}
}
